homeassistant/components/aurora_abb_powerone/__init__.py

```diff
@@ -10,40 +10,52 @@
 
 import logging
 
-from aurorapy.client import AuroraSerialClient
+from aurorapy.client import AuroraError, AuroraSerialClient
 
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import CONF_ADDRESS, CONF_PORT
 from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
 
-from .const import DOMAIN
+from .config_flow import validate_and_connect
+from .const import ATTR_SERIAL_NUMBER, DOMAIN
 
 PLATFORMS = ["sensor"]
 
 _LOGGER = logging.getLogger(__name__)
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Aurora ABB PowerOne from a config entry."""
 
     comport = entry.data[CONF_PORT]
     address = entry.data[CONF_ADDRESS]
-    serclient = AuroraSerialClient(address, comport, parity="N", timeout=1)
-
-    hass.data.setdefault(DOMAIN, {})[entry.unique_id] = serclient
-
+    ser_client = AuroraSerialClient(address, comport, parity="N", timeout=1)
+    # To handle yaml import attempts in darkeness, (re)try connecting only if
+    # unique_id not yet assigned.
+    if entry.unique_id is None:
+        try:
+            res = await hass.async_add_executor_job(
+                validate_and_connect, hass, entry.data
+            )
+            entry.unique_id = res[ATTR_SERIAL_NUMBER]
```

Comment:
> Please only wrap the line that can change in the try... except block.

Agreed. Done in #58887.

Reply:
> We need to use `hass.config_entries.async_update_entry` to change the unique id.
> 
> We also need to handle an existing entry with the same unique id. I think this could happen if we import the yaml without a unique id, then set up a user flow before the imported entry has been updated with a unique id. In that case we should probably remove the config entry without unique id, using `hass.config_entries.async_remove`.

Agreed. Done in #58887.